design/batch_retrieval_design.md

```diff
@@ -1,20 +1,378 @@
+| 4    | Conjur server is not up-to-date                              | - Outdated Conjur server is running<br />- Authenticator is defined<br />- Secrets defined in Conjur and mapped in K8s Secrets<br />- All mandatory values are defined in `custom-values.yaml`<br />- Host defined to Secrets Provider in `custom-values.yaml` does not have access to one of the Conjur secrets<br />- Customer installs Secrets provider Chart | Secrets Provider runs as a Job | - Secrets Provider pod authenticates and attempts to fetch Conjur secrets<br /><br />- Secrets Provider Job runs to exhaustion<br />- Application fails to be deployed<br />- Verify logs #1/4 appear | Automated        |
+
+## Logs
+
+|      | **Scenario**                                                 | **Log message**                                              | Log level |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | --------- |
+| 1    | Secrets Provider batch retrieval has partial success (Conjur) | Failed to retrieve a portion of the Conjur/DAP secrets: Reason: '%s' | Error     |
+| 2    | Secrets Provider batch retrieval total failure (Conjur)      | Failed to retrieve Conjur/DAP secrets: Reason: '%s'          | Error     |
+| 3    | Secrets Provider batch retrieval has partial success (Secrets Provider) | Skipped Kubernetes Secrets '%s'. Reason: Failed to retrieve DAP/Conjur secrets | Debug     |
+| 4    | Secrets Provider success                                     | Successfully updated '%d' out of '%d' Kubernetes Secrets     | Info      |
+| 5    | Old batch retrieval endpoint                                 | Warning: Secrets Provider cannot efficiently run because DAP/Conjur is not up to date. Please consider upgrading to the latest version | Warn      |
+
+## Performance tests
+
+@Abraham add here
+
+
+
+### Open questions
+
+1. Do we want this duplicated API documented? At this stage this API endpoint will be for internal use only.
+2. Do we ultimately want to move to a single Batch Request API and depricate the existing one?
+3. Do we want to support v4?
+
+### Future enhancements
+
+The following are ideas for possible extension of the batch retrieval endpoint API and not necessary directly related to the Secrets Provider project.
+
+1. Add secret versions as a filtering query parameter so we can get all secrets after a certain version
+2. Fetch all secrets that the user/host has permissions to
+3. Fetch all secrets after `/var/path/*` that the user/host has permissions to
+
+### Delivery Plan
+
+- [ ] Fix flaky tests *2 days*
+- [ ] Implement new API endpoint in Conjur *10 days*
```

Comment:
Shouldn't be more than 3 days.

Reply:
When I give estimates I don't only think of the work done on our end but also try to factor in cross-team work (review from our own team and from conjur-core, etc). I don't know if that is the best way to estimate but I find that we haven't been meeting our estimations because of that factor. 
I agree that this shouldn't be 10 but 3 seems a little too low. Maybe 5 or 6

Reply:
As discussed in a private channel we will split implementation and conjur-core reviews into two separate tickets